velox/type/Filter.cpp

```diff
@@ -458,6 +466,67 @@ bool MultiRange::testBytesRange(
   return false;
 }
 
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kMultiRange: {
+      const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+      bool bothNanAllowed = nanAllowed_ && multiRangeOther->nanAllowed_;
+      std::vector<std::unique_ptr<Filter>> merged;
+      for (auto const& filter : this->filters()) {
+        for (auto const& filterOther : multiRangeOther->filters()) {
+          auto innerMerged = filter->mergeWith(filterOther.get());
+          switch (innerMerged->kind()) {
+            case FilterKind::kAlwaysFalse:
+            case FilterKind::kIsNull:
+              break;
+            default:
+              merged.push_back(std::move(innerMerged));
+          }
+        }
+      }
+
+      if (merged.empty()) {
+        return nullOrFalse(bothNullAllowed);
+      } else if (merged.size() == 1) {
+        return merged[0]->clone(bothNullAllowed);
+      } else {
+        return std::unique_ptr<Filter>(
+            new MultiRange(std::move(merged), bothNullAllowed, bothNanAllowed));
+      }
+    }
+      /*
+       * Rules of MultiRange with IsNull/IsNotNull
+       * MultiRange(nullAllowed=true) AND IS NULL => IS NULL
+       * MultiRange(nullAllowed=true) AND IS NOT NULL =>
+       * MultiRange(nullAllowed=false) MultiRange(nullAllowed=false) AND IS NULL
+       * => ALWAYS FALSE MultiRange(nullAllowed=false) AND IS NOT NULL =>
+       * MultiRange(nullAllowed=false)
+       */
+    case FilterKind::kAlwaysTrue:
+    case FilterKind::kAlwaysFalse:
+    case FilterKind::kIsNull:
+      return other->mergeWith(this);
+    case FilterKind::kIsNotNull: {
+      if (this->nullAllowed_) {
+        std::unique_ptr<Filter> clonedFilter = this->clone();
+        std::unique_ptr<MultiRange> clonedMultiRrange(
+            static_cast<MultiRange*>(clonedFilter.release()));
+        std::vector<std::unique_ptr<Filter>> filters;
+        for (auto& filter : filters_) {
+          filters.push_back(filter->clone());
+        }
+        return std::make_unique<MultiRange>(
+            std::move(filters), /*nullAllowed=*/false, this->nanAllowed_);
```

Comment:
Perhaps, add MultiRange::clone(bool nullAllowed) and refactor this logic into `return this->close(false)`?